### ISSUE TITLE
macOS fails to detect if a path is a file or a directory on NFS if the file is not in current directory

### ISSUE THREAD

**Issue by illwieckz** [DaemonEngine/crunch#63]
This PR fixed provided an alternative code to detect if a path is a file or a directory to make it work on network file systems like NFS or SSHFS:

- https://github.com/DaemonEngine/crunch/pull/41

But on macOS with NFS the `stat` command seems to be broken, it only works if the file is in current directory.

Example, let's imagine your file is `/User/yourself/files/image.png`,

This works:

```sh
cd /User/yourself/files
crunch -file image.png -out image.crn
crunch -file /User/yourself/files/image.png -out image.crn
```

This doesn't work:

```sh
cd /User/yourself
crunch -file files/image.png -out image.crn
crunch -file /User/yourself/files/image.png -out image.crn
```

Even with the same full path it doesn't work if the file is not in the current directory, it will print:

```
Error: Cannot detect if given path is a file or a directory
Error: Failed finding files: /User/yourself/files/image.png
```

**Comment by illwieckz** [comment]
Even `open` returns `-1` on the file… That doesn't make sense.

**Comment by illwieckz** [comment]
The error reported by `perror` is: `No such file or directory`, this is wrong, the file exists.

**Comment by illwieckz** [comment]
> Even `open` returns `-1` on the file… That doesn't make sense.

What is also very wrong is that it doesn't work there, but it later works when reading the content of the file itself…

Anyway, I found a workaround, since `opendir` works, we can detect if a path is a directory, and then assume a path that is not a directory is a file. That assumption may be wrong but this implementation is only used as a fallback after two other implementations have failed.

**Comment by slipher** [comment]
I have a Mac machine. Can you give repro instructions for this?

**Comment by illwieckz** [comment]
In the following I'll name my Linux machine `linuxbox` and the macOS machine `macbox`.

I have a Linux machine with `/home/illwieckz/dev` folder, in which I store all my code repository, for some reasons my Unvanquished tree is in `/home/illwieckz/dev/buildme-unvanquished/Unvanquished`, so I have `/home/illwieckz/dev/buildme-unvanquished/Unvanquished/daemon/libs/crunch`. The user's uid/gid on Linux is `1000`.

On the Linux machine, the NFS `/etc/exports` contains:

```
/home/illwieckz/dev macbox(rw,sync,no_subtree_check,all_squash,anonuid=1000,anongid=1000) 
```

On macOS I mount the folder this way:

```sh
mkdir -pv ~/dev
sudo mount -t nfs -o resvport,noowners,locallocks,rw linuxbox:/home/illwieckz/dev /Users/illwieckz/dev
```

Then on macOS I go to `/Users/illwieckz/dev/buildme-unvanquished/Unvanquished/daemon/libs/crunch` and do:

```sh
mkdir build-macos
cd build-macos
cmake ..
make -j$(sysctl -n hw.logicalcpu)
```

Then I do:

```sh
./crunch -file ../test/unvanquished_64.png -out test.crn
```

It fails and says:

```
Error: Cannot detect if the given path is a file or a directory
Error: Failed finding files: /Users/illwieckz/dev/buildme-unvanquished/Unvanquished/daemon/libs/crunch/unvanquished_64.png
```

But it works if I do:

```sh
cp -a ../test/unvanquished_64.png .
./crunch -file unvanquished_64.png -out test.crn
```

It also works if I do:

```sh
cd ../test
../build-macos/crunch -file unvanquished_64.png -out ../build-macos/test.crn
```

**Comment by illwieckz** [comment]
I reproduce the exact same error on my FreeBSD machine with NFS.

The NFS export is:

```
/home/illwieckz/dev freebsdbox(rw,sync,no_subtree_check)
```

And I mount this way (users share the same id):

```
mount linuxbox:/home/illwieckz/dev /home/illwieckz/dev
```

I'll try with Linux.

**Comment by illwieckz** [comment]
Un fortunately I cannot reproduce the error on Linux:

```sh
sudo su -c "echo '/home/illwieckz/dev localhost(rw,sync,no_subtree_check)' >> /etc/exports"
sudo exportfs -r
mkdir -pv /home/illwieckz/dev2
sudo mount localhost:/home/illwieckz/dev /home/illwieckz/dev2
cd /home/illwieckz/dev2/buildme-unvanquished/Unvanquished/daemon/libs/crunch
mkdir -pv build-linux2
cd build-linux2
cmake ..
make -j$(nproc)
./crunch -file ../test/unvanquished_64.png -out test.crn
```
```
crunch: Advanced DXTn Texture Compressor (Daemon branch, Unity format variant)

crnlib version v1.04U 64-bit Built from git-4e24afa

Crunch is brought to you by:
- 2014-2024 Daemon Developers and contributors
  https://github.com/DaemonEngine/crunch
- 2017-2018 Alexander Suvorov and Unity Software Inc.
  https://github.com/Unity-Technologies/crunch/tree/unity
- 2010-2017 Richard Geldreich, Jr. and Binomial LLC and contributors
  https://github.com/BinomialLLC/crunch

Please report bugs here: https://github.com/DaemonEngine/crunch/issues

Reading source texture: "/home/illwieckz/dev2/buildme-unvanquished/Unvanquished/daemon/libs/crunch/test/unvanquished_64.png"
Texture successfully loaded in 0.001s
Source texture: 64x64, Levels: 1, Faces: 1, Format: A8R8G8B8
Apparent type: 2D map, Flags: R G B A Non-Flipped 
Generating mipmaps using filter "kaiser"
Generated 6 mipmap levels in 0.009s
Writing DXT5 texture to file: "test.crn"
Compressing using quality level 128
Processing: 100%              
Texture successfully written in 0.111s
Texture successfully processed in 0.120s
Input texture: 64x64, Levels: 1, Faces: 1, Format: A8R8G8B8
Input pixels: 4096, Input file size: 7471, Input bits/pixel: 14.592
Output texture: 64x64, Levels: 7, Faces: 1, Format: DXT5
Output pixels: 5461, Output file size: 2210, Output bits/pixel: 3.238

Total time: 0.121s
1 total file(s) successfully processed, 0 file(s) skipped, 0 file(s) failed.

Exit status: 0
```

But I reproduce the error with both FreeBSD and macOS NFS clients.

**Comment by slipher** [comment]
On my Mac box, running Crunch with an input file on NFS works OK with current master. I used the same options for the client mounting command: `sudo mount -t nfs -o resvport,noowners,locallocks,rw`. 

On the Debian VM with the NFS server it has this configuration:
```
$ sudo exportfs -v
/srv/nfs4     	192.168.1.138/32(ro,wdelay,root_squash,no_subtree_check,sec=sys,ro,secure,root_squash,no_all_squash)
```

**Comment by illwieckz** [comment]
But does it work if the input file is on NFS but on another folder than the current folder crunch is running?

**Comment by illwieckz** [comment]
I restarted my NFS server with those options, even remounted on macOS side, no difference on my end.

**Comment by slipher** [comment]
> But does it work if the input file is on NFS but on another folder than the current folder crunch is running?

I don't exactly understand the constraint, but it works if the PWD, the crunch build directory, and the input directory are all different, and the PWD is not in NFS.

**Comment by illwieckz** [comment]
Hmm, this doesn't work for me. I'm not running the latest macOS neither the latest FreeBSD so maybe it's a bug in a common component that has been updated since…

The constraint is that for example, you have crunch folder being on nfs and you do:

```
mkdir build
cd build
cmake ..
make -j$(sysctl -n hw.logicalcpu)
./crunch -file ../test/unvanquished_64.png -out test.crn
```

The current folder would be `crunch/build` but the input file would be in `crunch/test` folder. This is what doesn't work on my end, if I copy the file to `crunch/build` or if I change directory to `crunch/test` before running the command, it works.

If I read correctly your comment, you tested a case that would have reproduced the issue I face on my end.

**Comment by illwieckz** [comment]
To make the test complete, I copied my build folder to a non-NFS folder and when running crunch (stored in a non-NFS folder) on a file that is on NFS doesn't work on my end, unless I change directory to that NFS folder first.

**Comment by illwieckz** [comment]
I also tested changing directory to random place and calling crunch from there, it doesn't work for me if the file is elsewhere on NFS.

**Comment by illwieckz** [comment]
So, if it only affects some macOS/FreeBSD version, my workaround is maybe good enough™.

**Comment by slipher** [comment]
Can you reproduce such problems outside of Crunch? Like with a `stat` command or some small C program.

If problems occur with both Mac or FreeBSD clients, that suggests to me there is something wrong with the server. There could be other variables in the server-side filesystem configuration. Like some examples had mount points in the NFS serving directory; I did it as simple as possible putting a file directly in the NFS root.

**Comment by illwieckz** [comment]
> If problems occur with both Mac or FreeBSD clients, that suggests to me there is something wrong with the server.

macOS is based on FreeBSD and both that macOS and FreeBSD is from that same era. I don't get the error when the client is Linux while always using the same NFS server on both three systems.

> Can you reproduce such problems outside of Crunch? Like with a stat command or some small C program.

I haven't tried that, but something fishy is that crunch manages to open the file anyway, in another function.

And I first tried to implement an alternative implementation to `stat()` by doing `open()` then `fstat()`, and the code failed at `open()` call! This is nonsense because crunch manages to read the file later…

**Comment by illwieckz** [comment]
For completeness, these are the two other implementations I tried, they all fail:

```c++
    if (!guessed) {
      struct stat s;
      int fd = open(ep->d_name, O_RDONLY);
      if (fd != -1) {
        if (fstat(fd, &s) == 0) {
          is_directory = S_ISDIR(s.st_mode);
          is_file = S_ISREG(s.st_mode);
          guessed = true;
        }
      }
    }

    if (!guessed) {
      char cwd[PATH_MAX];
      if (getcwd(cwd, sizeof(cwd)) != NULL) {
        int fd = open(cwd, O_RDONLY);
        if (fd != -1) {
          struct stat s;
          if (fstatat(fd, ep->d_name, &s, 0) == 0) {
            is_directory = S_ISDIR(s.st_mode);
            is_file = S_ISREG(s.st_mode);
            guessed = true;
          }
        }
      }
    }
```

And the `fstat()` implementation fails at the `open()` step with the “No such file or directory” error.

**Comment by illwieckz** [comment]
In fact even my workaround doesn't work, I also get `No such file or directory` with `opendir()`, I probably made a mistake before that I fixed when beautifying my PoC prior to pushing the branch.

**Comment by illwieckz** [comment]
I modified the code to print `ep->d_name`, it gave me `build_a`, there is no file/folder with that name on the system, something is wrong elsewhere.

**Comment by illwieckz** [comment]
Wait, no. I actually have a `crunch/test/build-a` folder.

**Comment by illwieckz** [comment]
Well, the existing alternate code is broken by design, `ep->d_name` is the file name in the opened dir, and `stat()` looks for it in the current folder of course.

**Comment by illwieckz** [comment]
Ok, I have a proper fix, there was a bug in the code.

**Comment by illwieckz** [comment]
I have no idea how I got it working on Linux over NFS… But anyway, I now have a proper fix, it replaced my previous commit in #64:

- https://github.com/DaemonEngine/crunch/pull/64
